### PR TITLE
feat: Provide opt-in debug logging

### PR DESCRIPTION
**Pull request by dazuma**
Adds a `logger` attribute to each credential type (via BaseClient) and a pass-through from the high-level Credential class. The value is `nil` by default, but when set, certain events such as JWT application and OAuth2 token exchange are logged. Tokens are obfuscated via sha256.

### CODE REVIEW

**Review by aandreassa — APPROVED** [review]

File: lib/googleauth/external_account/base_credentials.rb [hunk 1]
```diff
@@ -129,18 +129,39 @@ def exchange_token
...
+        def log_token_request token_request
+          logger&.info do
+            Google::Logging::Message.from(
+              message: "Requesting access token from #{STS_GRANT_TYPE}",
+              "credentialsId" => object_id
+            )
+          end
+          logger&.debug do
+            digest = Digest::SHA256.hexdigest token_request[:subject_token].to_s
+            loggable_request = token_request.merge subject_token: "(sha256:#{digest})"
```
Comment (aandreassa) [inline lib/googleauth/external_account/base_credentials.rb:154]: Opt.: could be useful to add a test so we don't ever forget to replace the original values.

**Review by aandreassa — APPROVED** [review]

File: lib/googleauth/external_account/base_credentials.rb [hunk 2]
```diff
@@ -129,18 +129,39 @@ def exchange_token
...
-          )
+          }
+          log_token_request token_request
+          @sts_client.exchange_token token_request
+        end
+
+        def log_token_request token_request
+          logger&.info do
+            Google::Logging::Message.from(
+              message: "Requesting access token from #{STS_GRANT_TYPE}",
```
[the remark this answers is not in the document]
Comment (dazuma) [reply lib/googleauth/external_account/base_credentials.rb]: Sure.

File: lib/googleauth/external_account/base_credentials.rb (hunk 1 above)
Comment (dazuma) [reply lib/googleauth/external_account/base_credentials.rb:154]: It's actually not that important.

File: lib/googleauth/credentials.rb [hunk 3]
```diff
@@ -373,8 +377,8 @@ def initialize keyfile, options = {}
         else
           update_from_filepath keyfile, options
         end
+        setup_logging logger: options.fetch(:logger, :default)
         @project_id ||= CredentialsLoader.load_gcloud_project_id
-        @client.fetch_access_token! if @client.needs_access_token?
```
Comment (tartakynov) [inline lib/googleauth/credentials.rb:377]: was this change intentional?

Comment (dazuma) [reply lib/googleauth/credentials.rb:377]: @tartakynov Yes it was. Auth clients should be fetching tokens just-in-time when a request is actually made, rather than eagerly up front. Did this cause an issue?

Comment (tartakynov) [reply lib/googleauth/credentials.rb:377]: @dazuma yeah, I left a bug report here 👉  https://github.com/googleapis/google-auth-library-ruby/issues/508

Comment (Linuus) [reply lib/googleauth/credentials.rb:377]: @dazuma This change also broke our codebase... We now get this error when trying to call BigQuery:
```rb
Google::Cloud::UnauthenticatedError: Unauthorized
```

Upgrading to `1.12.2` etc does not help.
